### ISSUE TITLE
Action prediction with Yolo

### ISSUE THREAD

**Issue by lesreaper** [AlexeyAB/darknet#404]
This repository has come a long way since it was first forked. Great work!

I left Yolo and started working on other human action prediction models awhile ago now. Specifically, recognizing when someone leaves out a door. However, it has proved amazingly difficult to get models to generalize this one action in PyTorch or TF. 

I'm coming back and asking if Darknet and Yolo do something like this at a generally acceptable precision? The requirements would be for it to recognize a door, a person, a person disappearing through the door, and then a classification saying: Door Exited by Person

It may not even do it out of the box, I get that, but is it even possible to set up Action prediction with Yolo2 Object ID inferencing?

**Comment by AlexeyAB** [comment]
This is the simple classification task that looks like this, since 1:50: https://youtu.be/KxDs8G2EsvM?t=110
As you see this task can be solved even with XNOR-net - where used 1-bit values instead of float-32-bit: https://arxiv.org/pdf/1603.05279.pdf


![plato_xnor](https://cloud.githubusercontent.com/assets/4096485/22928510/e2b605c8-f2c7-11e6-8b50-1fb3bdaa32d5.jpg)


----

But if you want to detect actions on the video - then you can train and use Darknet Yolo v2 for this, because for Yolo already implemented detection on video. My recommendations, just:
* increase `#define FRAMES 15` or that minimum number of frames that the action takes - so that if the action has a high probability over a plurality of frames, then only in this case it is detected: https://github.com/AlexeyAB/darknet/blob/cd2bdec09030edf7da79ecdeb38d908c106850b3/src/demo.c#L18
* decrease `float nms = 0.01;` so that one detection is not divided into many: https://github.com/AlexeyAB/darknet/blob/cd2bdec09030edf7da79ecdeb38d908c106850b3/src/demo.c#L74
* train Yolo v2 where you should label actions instead of objects: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects
* as bonus - you will get coordinates of actions, this isn't required in your task, but in general it can be used if actions can occur in different places simultaneously

**Comment by lesreaper** [comment]
I understand extending out the frames, that makes sense.

In the video you showed, the action ID was easy as it was trained on a gun and a knife as objects, and could easily pick that object out of 1 frame.

However, action prediction does not exist on one frame. It only exists in the sequence of the 30 frames (~1 sec) of someone moving through an open door. Normally this is done with an LSTM or RNN where the output is fed into the input of the next frame.

I will give this a shot for sure. Maybe if I use the image of a person as they're half-hidden by a door, that will be enough for the custom object to be recognized.

Let me see how it goes!

Thanks!

**Comment by AlexeyAB** [comment]
Yes, in general you should use RNN / LSTM / GRU at the top of fully-convolutional-network. [RNN](https://github.com/AlexeyAB/darknet/blob/master/cfg/rnn.cfg) and [GRU](https://github.com/AlexeyAB/darknet/blob/master/cfg/gru.cfg) are in this repo, also LSTM was added to the original repo: https://github.com/pjreddie/darknet/pull/64

But to use RNN+Classifier (or RNN+Yolo) for detection of action - you should change source code of Darknet, and you should make a dataset of images grouped by actions - for each action should be a subset of images in a strict sequence.

It is required only if you want to detect difficult actions - with a strict sequence of several sub-actions. In your case this is required only:
* if you want to distinguish two actions: (1) the person left the room, and (2) when the person, started to go out from the room, but changed his mind and returned back
* or if you want to catch cases when a person enters the door with his back forward

RNN/LSTM/GRU are required in cases:
* if you want to detect sequences, to detect complex action that consist sequence of many of sub-actions - this is your case
* if you want to predict the advance of sequence in the future:
    * object tracking: https://github.com/Guanghan/ROLO
    * price prediction: https://lilianweng.github.io/lil-log/2017/07/08/predict-stock-prices-using-RNN-part-1.html
    * generation of text in the style of great writers: `darknet.exe rnn generate cfg/rnn.cfg tolstoy.weights -srand 2 -seed Chapter` where https://pjreddie.com/media/files/tolstoy.weights

----

Without RNN/LSTM/GRU the simplest way is to divide complex action into several simple sub-actions each of them has unique class_id and checking sequence of detected classes id in your hardcoded C-code.

**Comment by deimsdeutsch** [comment]
Is this updated to run on yolov3 tiny ?
